biorxiv/crawl_extract_metadata_csv.py

```diff
@@ -0,0 +1,155 @@
+        if 'Copyright' in label_div.text:
+            copyright_info_div = label_div.find_next_sibling('div', class_='field-items')
+            if copyright_info_div:
+                copyright_info = copyright_info_div.text.strip()
+            break
+
+    return [paper_title, doi, authors, posted, copyright_info]
+
+
+def crawl_and_extract_metadata(category, output_file="output_metadata.csv", max_pages=5, max_workers=10):
+    """ Crawls biorxiv site based on category and downloads the articles metadata in csv format
+    :param category: category to crawl, e.g., neuroscience
+    :param output_folder: output folder to save the downloaded pdf files
+    :param max_pages: pagination or the number of sub-pages pages to scan for
+    :param max_workers: Concurrency limit
+    :return: dict
+    """
+    current_page = 1
+
+    # Open the CSV file in append mode and write the header
+    with open(output_file, mode='a', newline='', encoding='utf-8') as file:
+        writer = csv.writer(file)
+
+        # Write the header only if the file is empty
+        if file.tell() == 0:
+            writer.writerow(['Paper Title', 'DOI', 'Authors', 'Posted', 'Copyright'])
+
+        with requests.Session() as session:
+            while current_page <= max_pages:
+                headers = {"User-Agent": get_random_user_agent()}
+                category_url = f"https://www.biorxiv.org/collection/{category}?page={current_page}"
+
+                try:
+                    response = session.get(category_url, headers=headers)
+                    response.raise_for_status()
+                except requests.RequestException as e:
+                    sys.stdout.write(f"Failed to retrieve page {current_page} for category: {category} due to: {e}")
+                    break
+
+                soup = BeautifulSoup(response.content, 'html.parser')
+                articles = soup.find_all('a', class_='highwire-cite-linked-title')
+
+                if not articles:
+                    sys.stdout.write("No more articles found.")
+                    break
+
+                sys.stdout.write(f"Page {current_page}: Found {len(articles)} articles in category '{category}'")
+
+                # Create a ThreadPoolExecutor
+                with ThreadPoolExecutor(max_workers=max_workers) as executor:
+                    # Submit all tasks and collect futures
+                    futures = [
+                        executor.submit(extract_article_info, urljoin("https://www.biorxiv.org", article['href'])) for
+                        article in articles]
+
+                    for future in as_completed(futures):
+                        result = future.result()
+                        if result:
+                            writer.writerow(result)
+                            sys.stdout.write(f"Extracted: {result[0]}")
+
+                current_page += 1
+                time.sleep(random.uniform(5, 10))  # Random delay between 5 and 10 seconds
```

Comment:
I understand you are using the sleep function so that biorxiv doesn't flag your script -- can this be shorter though? seems like a big bottleneck?

Reply:
I was using smaller one but then biorxiv was not responding. But I can test shorter time and update accordingly